### ISSUE TITLE
[Bug]: Panic on checkSuspendedStatefulsetsConformity 

### ISSUE THREAD

**Issue by apapia** [govirtuo/kube-ns-suspender#85]
### Version

v2.1.0

### What happened?

Thanks for this project.  It looks like it meets my use case exactly.  However, when I first tried to use this on my minikube setup, I hit a panic.  I'm on a slightly older version of Kubernetes.  Not sure if the issue is related to that:
```
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.14", GitCommit:"0f77da5bd4809927e15d1658fb4aa8f13ad890a5", GitTreeState:"clean", BuildDate:"2022-06-15T14:11:36Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

### Relevant log output

```bash
{"level":"debug","routine":"suspender","namespace":"cogito","step":"3/3 - handle desiredState","resource":"statefulsets","time":"2022-12-14T17:26:00+01:00","message":"checking suspended Conformity"}
{"level":"info","routine":"suspender","namespace":"cogito","statefulset":"keycloak","time":"2022-12-14T17:26:00+01:00","message":"scaling keycloak from 1 to 0 replicas"}
{"level":"info","routine":"suspender","namespace":"cogito","deployment":"compute-manager","time":"2022-12-14T17:26:00+01:00","message":"scaling compute-manager from 1 to 0 replicas"}
E1214 17:26:00.378244       1 runtime.go:78] Observed a panic: "assignment to entry in nil map" (assignment to entry in nil map)
goroutine 61 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x13287a0, 0x1661690})
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/runtime/runtime.go:74 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0001c0000})
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/runtime/runtime.go:48 +0x75
panic({0x13287a0, 0x1661690})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/govirtuo/kube-ns-suspender/engine.patchStatefulsetReplicas.func1()
	/build/engine/statefulset.go:61 +0x186
k8s.io/client-go/util/retry.OnError.func1()
	/go/pkg/mod/k8s.io/client-go@v0.22.2/util/retry/util.go:51 +0x33
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x40ce34, 0xc0000a1848})
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/wait/wait.go:217 +0x1b
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x168bb88, 0xc000046038}, 0xc0000a1950)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/wait/wait.go:230 +0x7c
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x13aa440)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/wait/wait.go:223 +0x39
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x40d187)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/wait/wait.go:418 +0x5f
k8s.io/client-go/util/retry.OnError({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x15600d0, 0xc000604840)
	/go/pkg/mod/k8s.io/client-go@v0.22.2/util/retry/util.go:50 +0xf1
k8s.io/client-go/util/retry.RetryOnConflict(...)
	/go/pkg/mod/k8s.io/client-go@v0.22.2/util/retry/util.go:104
github.com/govirtuo/kube-ns-suspender/engine.patchStatefulsetReplicas({0x168bb88, 0xc000046030}, 0xc0003af4a0, {0xc00048acf0, 0x6}, {0xc0002de930, 0x8}, {0x14b63de, 0x12}, 0x0)
	/build/engine/statefulset.go:53 +0x1d2
github.com/govirtuo/kube-ns-suspender/engine.checkSuspendedStatefulsetsConformity({0x168bb88, 0xc000046030}, {{0x167daa8, 0xc00038bee0}, 0x0, {0x0, 0x0}, {0xc00042e600, 0x2b, 0x1f4}, ...}, ...)
	/build/engine/statefulset.go:43 +0x1f1
github.com/govirtuo/kube-ns-suspender/engine.(*Engine).Suspender.func6()
	/build/engine/suspender.go:266 +0xa8
created by github.com/govirtuo/kube-ns-suspender/engine.(*Engine).Suspender
	/build/engine/suspender.go:265 +0x3a05
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map

goroutine 61 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0001c0000})
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x13287a0, 0x1661690})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/govirtuo/kube-ns-suspender/engine.patchStatefulsetReplicas.func1()
	/build/engine/statefulset.go:61 +0x186
k8s.io/client-go/util/retry.OnError.func1()
	/go/pkg/mod/k8s.io/client-go@v0.22.2/util/retry/util.go:51 +0x33
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x40ce34, 0xc0000a1848})
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/wait/wait.go:217 +0x1b
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x168bb88, 0xc000046038}, 0xc0000a1950)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/wait/wait.go:230 +0x7c
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x13aa440)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/wait/wait.go:223 +0x39
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x40d187)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.2/pkg/util/wait/wait.go:418 +0x5f
k8s.io/client-go/util/retry.OnError({0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0}, 0x15600d0, 0xc000604840)
	/go/pkg/mod/k8s.io/client-go@v0.22.2/util/retry/util.go:50 +0xf1
k8s.io/client-go/util/retry.RetryOnConflict(...)
	/go/pkg/mod/k8s.io/client-go@v0.22.2/util/retry/util.go:104
github.com/govirtuo/kube-ns-suspender/engine.patchStatefulsetReplicas({0x168bb88, 0xc000046030}, 0xc0003af4a0, {0xc00048acf0, 0x6}, {0xc0002de930, 0x8}, {0x14b63de, 0x12}, 0x0)
	/build/engine/statefulset.go:53 +0x1d2
github.com/govirtuo/kube-ns-suspender/engine.checkSuspendedStatefulsetsConformity({0x168bb88, 0xc000046030}, {{0x167daa8, 0xc00038bee0}, 0x0, {0x0, 0x0}, {0xc00042e600, 0x2b, 0x1f4}, ...}, ...)
	/build/engine/statefulset.go:43 +0x1f1
github.com/govirtuo/kube-ns-suspender/engine.(*Engine).Suspender.func6()
	/build/engine/suspender.go:266 +0xa8
created by github.com/govirtuo/kube-ns-suspender/engine.(*Engine).Suspender
	/build/engine/suspender.go:265 +0x3a05
```

**Comment by eze-kiel** [comment]
Hi @apapia, thanks for reporting this issue! At first glance, this seems to be a purely software-related issue, nothing to do with your Kubernetes version.

The error `assignment to entry in nil map` occurs when trying to write to an uninitialized map ([PoC](https://playground.hugoblanc.com/snippet/KZz2lIRE8XR)).

I'll try to have a look as soon as possible and keep you posted!

**Comment by eze-kiel** [comment]
Did you started `kube-ns-suspender` with the metric server enabled? As the only `map` used is used for the metrics

**Comment by apapia** [comment]
Thanks for the quick response!  I put together a quick helm chart and I can share the generated files.

```
Loaded config file: binnacles/kube-ns-suspender.yaml
---
# Source: kube-ns-suspender/templates/serviceaccount.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    helm.sh/chart: kube-ns-suspender-0.0.1
    app.kubernetes.io/name: kube-ns-suspender
    app.kubernetes.io/instance: kube-ns-suspender
    app.kubernetes.io/version: "1.0"
    app.kubernetes.io/managed-by: Helm
  name: kube-ns-suspender
---
# Source: kube-ns-suspender/templates/cluster-role.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: kube-ns-suspender
  labels:
    app: kube-ns-suspender
rules:
- apiGroups:
  - ""
  resources:
  - namespaces
  verbs:
  - get
  - list
  - update
- apiGroups:
  - apps
  resources:
  - deployments
  - deployments/scale
  - statefulsets
  verbs:
  - get
  - list
  - update
- apiGroups:
  - batch
  resources:
  - cronjobs
  verbs:
  - get
  - list
  - update
---
# Source: kube-ns-suspender/templates/cluster-role-binding.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kube-ns-suspender
  labels:
    helm.sh/chart: kube-ns-suspender-0.0.1
    app.kubernetes.io/name: kube-ns-suspender
    app.kubernetes.io/instance: kube-ns-suspender
    app.kubernetes.io/version: "1.0"
    app.kubernetes.io/managed-by: Helm
subjects:
- kind: ServiceAccount
  name: kube-ns-suspender
  namespace: kube-ns-suspender
roleRef:
  kind: ClusterRole
  name: kube-ns-suspender
  apiGroup: rbac.authorization.k8s.io
---
# Source: kube-ns-suspender/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: kube-ns-suspender
  labels:
    helm.sh/chart: kube-ns-suspender-0.0.1
    app.kubernetes.io/name: kube-ns-suspender
    app.kubernetes.io/instance: kube-ns-suspender
    app.kubernetes.io/version: "1.0"
    app.kubernetes.io/managed-by: Helm
spec:
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: http
    - name: http-metrics
      port: 8444
      protocol: TCP
      targetPort: http-metrics
  selector:
    app.kubernetes.io/name: kube-ns-suspender
    app.kubernetes.io/instance: kube-ns-suspender
  type: ClusterIP
---
# Source: kube-ns-suspender/templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    helm.sh/chart: kube-ns-suspender-0.0.1
    app.kubernetes.io/name: kube-ns-suspender
    app.kubernetes.io/instance: kube-ns-suspender
    app.kubernetes.io/version: "1.0"
    app.kubernetes.io/managed-by: Helm
  name: kube-ns-suspender
spec:
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/name: kube-ns-suspender
      app.kubernetes.io/instance: kube-ns-suspender
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        checksum/configmap: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
      labels:
        app: kube-ns-suspender
        version: v2.1.0
        app.kubernetes.io/name: kube-ns-suspender
        app.kubernetes.io/instance: kube-ns-suspender
    spec:
      containers:
        - name: kube-ns-suspender
          env:
            - name: KUBE_NS_SUSPENDER_UI_EMBEDDED
              value: "true"
          image: "ghcr.io/govirtuo/kube-ns-suspender:v2.1.0"
          imagePullPolicy: IfNotPresent
          livenessProbe:
            httpGet:
              path: /
              port: http
          ports:
            - containerPort: 8080
              name: http
            - containerPort: 2112
              name: http-metrics
            - containerPort: 4455
              name: http-pprof
          readinessProbe:
            httpGet:
              path: /
              port: http
          resources:
            limits:
              cpu: 100m
              memory: 128Mi
            requests:
              cpu: 100m
              memory: 128Mi
          startupProbe:
            failureThreshold: 10
            httpGet:
              path: /
              port: http
          volumeMounts:
      dnsPolicy: ClusterFirst
      hostNetwork: false
      imagePullSecrets:
        - name: docker-cogitocorp-us-regcred
      serviceAccount: kube-ns-suspender
      terminationGracePeriodSeconds: 30
      volumes:
---
# Source: kube-ns-suspender/templates/ingress.yaml
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  annotations:
    konghq.com/https-redirect-status-code: "302"
    konghq.com/preserve-host: "false"
    konghq.com/protocols: https
    kubernetes.io/ingress.class: kong
  labels:
    helm.sh/chart: kube-ns-suspender-0.0.1
    app.kubernetes.io/name: kube-ns-suspender
    app.kubernetes.io/instance: kube-ns-suspender
    app.kubernetes.io/version: "1.0"
    app.kubernetes.io/managed-by: Helm
  name: kube-ns-suspender
spec:
  rules:
    - host: kube-ns-suspender.mini
      http:
        paths:
          - path: /
            backend:
              serviceName: kube-ns-suspender
              servicePort: http
---
# Source: kube-ns-suspender/templates/tests/test-connection.yaml
apiVersion: v1
kind: Pod
metadata:
  name: "kube-ns-suspender-test-connection"
  labels:
    helm.sh/chart: kube-ns-suspender-0.0.1
    app.kubernetes.io/name: kube-ns-suspender
    app.kubernetes.io/instance: kube-ns-suspender
    app.kubernetes.io/version: "1.0"
    app.kubernetes.io/managed-by: Helm
  annotations:
    "helm.sh/hook": test-success
spec:
  containers:
    - name: wget
      image: busybox
      command: ['wget']
      args:  ['kube-ns-suspender:80']
  restartPolicy: Never

```

**Comment by eze-kiel** [comment]
Okay, I think I reproduced the bug successfully!

## Steps to reproduce

1. Start minikube with the same Kubernetes version:
```
$ minikube start  --kubernetes-version=v1.21.0
```

2. Deploy `kube-ns-suspender` and the testing manifests (under `manifests/testing-namespace`).

3. Add the annotations to the new namespace:

```diff
kind: Namespace
metadata:
  annotations:
+    kube-ns-suspender/controllerName: kube-ns-suspender
+    kube-ns-suspender/dailySuspendTime: 6:22PM
+    kube-ns-suspender/desiredState: Suspended
```

with a `dailySuspendTime` that is in the future (1min ahead)

4. Delete all the annotations on the StatefulSet

5. Enjoy

![image](https://user-images.githubusercontent.com/25285375/207666111-ecdf6c97-3df5-4b40-a319-1f2b6967b83d.png)

## Probable explaination

The "faulty" code is likely to be:

```go
if repl == 0 {
	result.Annotations[prefix+originalReplicas] = strconv.Itoa(int(*result.Spec.Replicas))
}
```

in `engine/statefulset.go` on line 61.

This tries to update annotations on the StatefulSet object. Those annotations are under the form of a map. So when it gets the annotations, the map is not initialized, it will fail. 

@apapia can you check that the StatefulSets you're have an annotation block defined? The command is likely to be something like:

```
$ kubectl get -n <namespace> statefulsets.apps <name> -o yaml
```

Thanks!

**Comment by apapia** [comment]
I added some debug logging and traced the issue to the same line of code.  Indeed, it is failing on a StatefulSet that is created by an Operator and does not have an Annotation block when I run the suggested command.  I believe that kube-ns-suspender should handle this case though.  Otherwise, I need to modify the Operator code to add an empty Annotation block.

**Comment by eze-kiel** [comment]
Usually, some annotations are added by Kubernetes on objects (`kubectl.kubernetes.io/last-applied-configuration` etc.), but you're right, this case should be handled by kube-ns-suspender. I'll work on it and keep you updated!

**Comment by apapia** [comment]
I added these lines and it is working for me now:
```
                if repl == 0 {
+                       if result.Annotations == nil {
+                               result.Annotations = make(map[string]string)
+                       }
                        result.Annotations[prefix+originalReplicas] = strconv.Itoa(int(*result.Spec.Replicas))
```

Let me know if you want me to Pull Request this change.  I imagine you can do it faster though.

**Comment by eze-kiel** [comment]
That's exactly the change I'm testing.
Feel free to open a PR if you want to contribute, otherwise I'll do it :wink:

**Comment by eze-kiel** [comment]
Fixed by #86
